hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/SparkFileFormatInternalRowReaderContext.scala

```diff
@@ -99,4 +104,49 @@ class SparkFileFormatInternalRowReaderContext(baseFileReader: Option[Partitioned
     })
     deserializer.deserialize(avroRecord).get.asInstanceOf[InternalRow]
   }
+
+  override def mergeBootstrapReaders(skeletonFileIterator: ClosableIterator[InternalRow],
+                                     dataFileIterator: ClosableIterator[InternalRow]): ClosableIterator[InternalRow] = {
+    doBootstrapMerge(skeletonFileIterator.asInstanceOf[ClosableIterator[Any]],
+      dataFileIterator.asInstanceOf[ClosableIterator[Any]])
+  }
+
+
+  protected def doBootstrapMerge(skeletonFileIterator: ClosableIterator[Any], dataFileIterator: ClosableIterator[Any]): ClosableIterator[InternalRow] = {
+    new ClosableIterator[Any] {
+      val combinedRow = new JoinedRow()
+
+      override def hasNext: Boolean = {
+        //If the iterators are out of sync it is probably due to filter pushdown
+        checkState(dataFileIterator.hasNext == skeletonFileIterator.hasNext,
+          "Bootstrap data-file iterator and skeleton-file iterator have to be in-sync!")
+        dataFileIterator.hasNext && skeletonFileIterator.hasNext
+      }
+
+      override def next(): Any = {
```

Comment:
This reminds me: do we need to support vectorization for other query types explicitly?